File: c/grains/grains.c
#include "grains.h"

uint64_t square(uint8_t index)
{
    uint64_t    result;

    result = 1;
    if (index < 1 || index > 64)
    {
        return (0);
    }
    while (index > 1)
    {
        result *= 2;
        index--;
    }
    return (result);
}
uint64_t total(void)
{
    uint64_t    result;
    int         i;

    result = 0;
    i = 1;
    while (i <= 64)
    {
        result += square(i);
        i++;
    }
    return (result);
}

File: go/grains/grains.go
// Package grains counts grains of wheat on chessboard.
package grains

import (
	"fmt"
)

// Square returns number of wheat grains on certain chess square.
func Square(squareNo int) (uint64, error) {
	if squareNo < 1 || squareNo > 64 {
		return 0, fmt.Errorf("bad square number %d, must be between 1 and 64", squareNo)
	}
	return 1 << (squareNo - 1), nil
}

// Total returns number of wheat grains on all chess squares together.
func Total() uint64 {
	return 1<<64 - 1
}
